clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -794,45 +935,224 @@ scheduler_tmpl_t<RecordType, ReaderType>::advance_region_of_interest(
+                            segment.start_instruction &&
+                        // When we skip our separator+timestamp markers are at the
+                        // prior instr ord so do not wait for that.
+                        outputs_[output].record[outputs_[output].record_index].type !=
+                            schedule_record_t::SKIP) {
+                        // Some other output stream has not advanced far enough, and we do
+                        // not support multiple positions in one input stream: we wait.
+                        // XXX i#5843: We may want to provide a kernel-mediated wait
+                        // feature so a multi-threaded simulator doesn't have to do a
+                        // spinning poll loop.
+                        VPRINT(this, 3,
+                               "next_record[%d]: waiting for input %d instr #%" PRId64
+                               "\n",
+                               output, index, segment.start_instruction);
+                        // Give up this input and go into a wait state.
+                        // We'll come back here on the next next_record() call.
+                        set_cur_input(output, INVALID_INPUT_ORDINAL);
+                        outputs_[output].waiting = true;
+                        return sched_type_t::STATUS_WAIT;
+                    }
+                }
+                // Also wait if this segment is ahead of the next-up segment on another
+                // output.  We only have a timestamp per context switch so we can't
+                // enforce finer-grained timing replay.
+                for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); i++) {
+                    if (i != output &&
+                        outputs_[i].record_index + 1 <
+                            static_cast<int>(outputs_[i].record.size()) &&
+                        segment.timestamp >
+                            outputs_[i].record[outputs_[i].record_index + 1].timestamp) {
+                        VPRINT(this, 3,
+                               "next_record[%d]: waiting because timestamp %" PRIu64
+                               " is ahead of output %d\n",
+                               output, segment.timestamp, i);
+                        // Give up this input and go into a wait state.
+                        // We'll come back here on the next next_record() call.
+                        set_cur_input(output, INVALID_INPUT_ORDINAL);
+                        outputs_[output].waiting = true;
+                        return sched_type_t::STATUS_WAIT;
+                    }
+                }
+                if (segment.type == schedule_record_t::SKIP) {
+                    std::lock_guard<std::mutex> lock(*inputs_[index].lock);
+                    if (segment.stop_instruction == 0) {
+                        // A stop of 0 means an artificial end to this input as we're past
```

Comment:
Is it worth creating a new schedule_record_t value to denote need to add thread exit?

Reply:
Done.